### PR TITLE
arch-riscv: This commit fixes bug in vfmv.f.s impl. in riscv

### PR DESCRIPTION
**Pull request by debjyoti0891**
The existing implementation of vfmv instruction did not type cast the first element of the source vector, which caused the "freg" to interpret the result as a NaN.

With the type cast to f32, the value is correctly recognized as float and sign extended to be stored in the fd register.

Git issue: https://github.com/gem5/gem5/issues/827

Change-Id: Ibe9873910827594c0ec11cb51ac0438428c3b54e

### CODE REVIEW

**Review by hnpl — COMMENTED** [review]
Hi, thanks for the fix.

I wonder if we should always cast the value to float32. My understanding is that the values in a vector register could be a set of float16 values or a set of float64 values too. Maybe we need to obtain the type from VTYPE.

~However, my understanding of vfmv.v.s is that it should be floating point type agnostic as we are merely copying bits from 1 floating point reg to a vector reg.~

According to https://github.com/riscv/riscv-v-spec/issues/339, we'll need to check for NaN if SEW != FLEN.

**Comment by debjyoti0891** [comment]
Hi @hnpl ,

Thanks for the feedback. Indeed, the cast should be based on the SEW. I am curious if I could use the [spike reference implementation](https://github.com/riscv-software-src/riscv-isa-sim/blob/master/riscv/insns/vfmv_s_f.h) as a baseline.

For NaN box, I am curious about the interpretation. 
- Do I check if the value is NaN in element 0 for each datatype (fp8, fp16, fp32,fp64)? 
- If this is true, the result stored in the floating point register is a canonical nan (7fc00000).

File: src/arch/riscv/isa/decoder.isa [hunk 1]
```diff
@@ -2746,7 +2746,7 @@ decode QUADRANT default Unknown::unknown() {
                         // The encodings corresponding to the masked versions
                         // (vm=0) of vfmv.f.s are reserved
                         0x1: VectorNonSplitFormat::vfmv_f_s({{
-                            freg_t fd = freg(Vs2_vu[0]);
+                            freg_t fd = freg(f32(Vs2_vu[0]));
```
Comment (rogerchang23424) [inline src/arch/riscv/isa/decoder.isa]: From `template/vector_aarith.isa`

```
def template VectorFloatNonSplitDecodeBlock {{

switch(machInst.vtype8.vsew) {
case 0b010: return new %(class_name)s<float32_t>(machInst);
case 0b011: return new %(class_name)s<float64_t>(machInst);
default: GEM5_UNREACHABLE;
}
```

I think we also need to check the type of `vu` before casting the `f32` and `f64`. We can assume the `FLEN = 64` as we don't have `FLEN` configuration.

**Comment by ivanaamit** [comment]
@debjyoti0891, can you address Roger's comment (https://github.com/gem5/gem5/pull/863#discussion_r1488792013)? We would like to merge this PR. Thank you.

**Comment by debjyoti0891** [comment]
> @debjyoti0891, can you address Roger's comment ([#863 (comment)](https://github.com/gem5/gem5/pull/863#discussion_r1488792013))? We would like to merge this PR. Thank you.

surely, I plan to wrap this up in the next week.

**Comment by ivanaamit** [comment]
@debjyoti0891, a friendly reminder to address pending comment. Thanks.

**Review by rogerchang23424 — APPROVED** [review]
Thanks for updating

**Review by powerjg — APPROVED** [review]

**Comment by debjyoti0891** [comment]
@ivanaamit , a friendly reminder to merge the PR

**Pull request merged**